Enable authentication by user impersonation and writing updating and deleting time series and data points. General improvements and a few error fixes.

tovop opened the pull request:
## Summary

### Added
- Support for authentication by user impersonation
- Support for paginated response from the web API
- Methods for writing, updating and deleting time series and data points
- Unit tests
- Notebook example on creating, updating and deleting time series and data points

### Changed
- `DataPoints.last` to `DataPoints.latest` to match the web API
- The client now used version 1.5 of the Omnia Timeseries web API
- Removed `TimeSeries.count` method as there were no end-point for that in the web API

### Fixed
Converting strings in nested dictionaries from/to snake-case and camel-case is now working properly.

Resolves #2 #3 #4

tovop committed:
Fetch paginated query results.

tovop committed:
Fix camel-cased dump of resource data.

tovop committed:
Rename `DataPoints.last` to `DataPoints.latest` to harmonize with API.

tovop committed:
Improve handling of paginated response and use of the `limit` parameter.

tovop committed:
Use API version 1.5.

tovop committed:
Clean __init__ file.

tovop committed:
Remove `count` method on `TimeSeries` resources. Not web API endpoint
for it.

Check possible aggregates like SQL COUNT.

tovop committed:
Improve conversion of snake-cased and camel-cased data.

tovop committed:
Improve test coverage.

tovop committed:
Update README.

tovop committed:
Merge branch 'master' into add-tests

tovop committed:
Update README.

tovop committed:
Merge branch 'master' into add-tests

tovop committed:
Merge remote-tracking branch 'origin/add-tests' into add-tests

tovop committed:
Update build recipe.

tovop committed:
Update build recipe.

tovop commented:
@ssorenes I took the liberty to request a review from you. The most important item in this PR is the suite of unittests.

tovop committed:
Add authentication by user impersonation.

Fixes #2

tovop committed:
Update README.

tovop committed:
Make test more robust.

Sometimes the API call is slow and the datetime interval check fails.

tovop committed:
Update description of `limit` parameter in docstring accordin to
changes from v1.3 to v1.5 of the API.

tovop committed:
Add new method `_patch` on `OmniaClient`.

tovop committed:
Make methods: `get`, `post`, `patch`, `put` and `delete` public.

tovop committed:
Improve handling of lowerCamelCase URL parameters.

tovop committed:
New method `create` to create single timeseries object.

tovop committed:
Catch message from API whenever no data is returned.

tovop committed:
Corrected media type for POST queries.

tovop committed:
Correct the sequence of url segments for DELETE queries.

tovop committed:
Create custom exceptions.

tovop committed:
Add end point for deleting time series..

tovop committed:
Add test case for `create` and `delete` end points.

tovop committed:
Catch empty response to avoid triggering IndexErrors etc.

tovop committed:
Add methods `add_data` and `delete_data` to the time_series API.

tovop committed:
Add `add_data` and `delete_data` methods to `TimeSeries` and `TimeSeriesList` resources
and `update` method to `DataPoint`, `DataPoints` and `DataPointsList`
resources.

tovop committed:
Create new test cases for adding and removing time series and data points.

tovop committed:
Ensure omnia datetime format is used in `TimeSeriesAPI.data`.

tovop committed:
Create tests for creation and deletion of timeseries.

tovop committed:
Create API end point for updating a time series. Add tests.

tovop committed:
Create `update` method on `TimeSeries` resource.

tovop committed:
Add notebook example on creating and updating time series and data points.

tovop commented:
@Timothy-Edward-Kendon, it is ready for review now.

tovop committed:
Add tests for creation and deletion of data points.

epa095 commented on tests/test_apiclient.py:
```diff
@@ -5,7 +5,68 @@
-class CreateAndDeleteDataPointsTestCase(unittest.TestCase):
+# TODO: Activate this test case once the responsiveness of the Timeseries API has improved. Now it takes too long from
+#  the query is executed to the datapoint is retrievable from Omnia. Also the buggy 500 error is problematic.
+# class DataPointsTestCase(unittest.TestCase):
```
This looks very much like a integration test, not a unit test;-) It is possible to mark the test like e.g.
```python
@unittest.skipIf(
    os.getenv("OMNIA_TIMESERIES_API_AUTH_KEY") is None,
    reason="Skipping test, OMNIA_TIMESERIES_API_AUTH_KEY not set in environment variable",
)
```
Then it will skip the test unless the auth-key is specified, but you can still keep the test and run it "on demand".

tovop replied:
I agree, it is an integration test. I did not think of that use case, I have been very focused on ensuring that the integration works.

Do you know, if this can be achieved using `Pytest` with test fixtures and separating unit test and function/integration tests? Or is your suggestion a better approach?

epa095 replied:
Yeah, one can do exactly the same with [pytest](http://doc.pytest.org/en/latest/skipping.html#id1). You can also just mark the tests using e.g. `@pytest.mark.integrationtest`, and then set the `setup.cfg` what should be the default testing when people write `python setup.py test`. [This](https://github.com/equinor/gordo/blob/4793dff1e7dac85790f4c438a901ce307eb4a71f/setup.cfg#L4) is how we are doing it, where we run everything except the dockertest by default. And then we put all the "global" pytest configuration into https://github.com/equinor/gordo/blob/master/pytest.ini

tovop replied:
Thanks, interesting example. I will look closer to to that pytest setup. Gordo might also become relevant for our project.

epa095 commented on omnia_timeseries_sdk/client.py:
```diff
@@ -47,28 +51,36 @@ def __init__(self, base_url: str = None, log_level: str = None):
+        # service-to-service / machine-to-machine authentication using a shared secret requires generally different
+        # resource and client ids and a shared secret (client secret). See also
+        # https://github.com/equinor/OmniaPlant/wiki/Authentication-&-Authorization
+        resource_id = os.getenv("omniaResourceId", "141369bd-3dca-4b55-825b-56ad4a69b1fc")
```
Are you sure these can be online? Same holds for the tenant ID. I have heard conflicting information about wether it is OK or not.

tovop replied:
First of all, thanks you for taking the time reviewing and proposing changes.

When I started this project I asked around and never really got a clear answer.  Considering that resource and client IDs are publicly available on https://github.com/equinor/OmniaPlant/wiki/Authentication-&-Authorization and the fact that it is useless without a shared secret or an equinor user with the necessary privileges, I decided to move on with these defaults.

Do you agree?

epa095 replied:
Should probably be ok then. I also found that we have exposed the tenant ID in public github issues, so I hope its fine :-p

epa095 commented on omnia_timeseries_sdk/client.py:
```diff
@@ -142,47 +124,114 @@ def _do_request(self, method: str, resource: str, version: str, endpoint: str, p
         logging.debug(msg)
 
         try:
             connection = http.client.HTTPSConnection(self.base_url)
```
Is there any reason not to use the `request` package? It is kind of the default choice for these things, to the point that even the documentation for `http.client.HTTPSConnection` sais  
> The Requests package is recommended for a higher-level HTTP client interface. 

With it you [wont have to craft the query_url manually](https://requests.readthedocs.io/en/master/user/quickstart/#passing-parameters-in-urls)

tovop replied:
Not a very good reason. I started this project with [this](https://api.equinor.com/docs/services/Timeseries-api-v1-5/operations/getTimeseries?) sample Python code and I have not prioritized to replace it with the `request` package. But that has been in the back of my mind all the way because I, as you, favour `request`.

tovop replied:
I will improve this once when I get time. See #6

tovop committed:
PEP8 improvement.

tovop committed:
Create configuration classes to enable separating between test and prod
environments in an easy way.

tovop committed:
Create Config classes to ease configuration.

tovop committed:
Initialize datapoints with id, name and unit also.

tovop committed:
Avoid error when created_time and changed_time are None.

tovop committed:
Adjust url to allow for plant-beta and plant environments.

tovop committed:
Direct integration tests towards the `plant-beta` environment instead of the production env.

- Skip integration tests if no credentials.
- Use custom dummy time series and data points for testing, not "real" data.

tovop committed:
Chande default value for parameter `step` on TimeSeriesResource.

tovop committed:
Use pytest for managing unit tests.

tovop committed:
Use pytest for managing integration tests.

tovop committed:
Use pytest for managing integration tests.

tovop committed:
Fix broken integration tests.

Were missing necessary fixtures.

tovop committed:
Remove unused imports.

tovop committed:
Disable 2 integration tests that fail due to latency.

It takes too long time from time series are created to these are discoverable through the `list` end point (GET /timeseries/v1.5/). This is a known issue and may be improved in API v1.6.

Timothy-Edward-Kendon reviewed (approved):
Approved. Excellent
